state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,214 @@
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use serde::{Deserialize, Serialize};
+use sp_std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
+
+#[cfg(feature = "runtime-benchmarks")]
+use cf_chains::benchmarking_value::BenchmarkValue;
+
+use crate::{
+	electoral_system::{
+		AuthorityVoteOf, ConsensusVotes, ElectionReadAccess, ElectionWriteAccess, ElectoralSystem,
+		ElectoralWriteAccess, VotePropertiesOf,
+	},
+	vote_storage::{self, VoteStorage},
+	CorruptStorageError, ElectionIdentifier,
+};
+use cf_chains::sol::{
+	api::ContractSwapAccountAndSender, MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES,
+	MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS,
+	NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES,
+};
+use cf_utilities::success_threshold_from_share_count;
+use frame_support::{
+	pallet_prelude::{MaybeSerializeDeserialize, Member},
+	sp_runtime::traits::CheckedSub,
+	Parameter,
+};
+use itertools::Itertools;
+use sp_std::vec::Vec;
+
+pub trait SolanaVaultSwapAccountsHook<Account, SwapDetails, E> {
+	fn close_accounts(accounts: Vec<Account>) -> Result<(), E>;
+	fn initiate_vault_swap(swap_details: SwapDetails);
+	fn get_number_of_available_sol_nonce_accounts() -> usize;
+}
```

Comment:
I don't think we should be coupling this so tightly to Solana. 

Can we not follow the same patterns/conventions as we previously have, with the hooks  defined externally rather than in the pallet?

Reply:
What I mean is that this electoral system includes a lot of non-election/consensus related logic inside of itself. ie. the interfaces for other electoral systems is defined in term of hooks that trigger externally-defined logic. However here the 'hooks' are not really hooks: they are accessors to specific external functionality, and the logic for how to process those things is internal to the election. We're mixing contexts.

Rather than `get_number_of_available_sol_nonce_accounts` and `initiate_vault_swap`, I think it should just have a single hook that says something like `on_account_change(added, removed)`; and the logic of *what* to do with that information should be defined externally.

Reply:
Given that its a custom election which is not intended to be re-used, I dont see why we cant have some custom logic in it especially if that logic is closely tied to what the election is about. We can rename the trait to not indicate that it is a hook but a general trait that calls some external code. The idea of injecting a trait that provides external code access is general enough here and the ElectoralSystem trait doesnt force us that the trait injected necessarily has to fit the definition of a hook (whatever that might be).

Even if we move out some of the functionality we would then need that logic to return some information back to the electoral system which it then uses to create a new election which would still mean it would still not function like a hook. 

In other electoral systems as well the code in on_finalize of the ElectoralSystem includes logic of first checking for consensus (calling the check_consensus) and then executing some logic to then decide whether to start a new election and with what properties. It is also the same here with just some external logic accessed through the Hook.